core/server/api/upload.js

```diff
@@ -23,27 +24,42 @@ upload = {
      * @returns {Promise} Success
      */
     add: function (options) {
-        var store = storage.getStorage(),
-            filepath;
+        var tasks,
+            attrs = ['uploadimage'];
 
-        // Check if a file was provided
-        if (!utils.checkFileExists(options, 'uploadimage')) {
-            return Promise.reject(new errors.NoPermissionError(i18n.t('errors.api.upload.pleaseSelectImage')));
+        function validate(options) {
+            // Check if a file was provided
+            if (!utils.checkFileExists(options.data, 'uploadimage')) {
+                return Promise.reject(new errors.NoPermissionError(i18n.t('errors.api.upload.pleaseSelectImage')));
+            }
+
+            // Check if the file is valid
+            if (!utils.checkFileIsValid(options.data.uploadimage, config.uploads.contentTypes, config.uploads.extensions)) {
+                return Promise.reject(new errors.UnsupportedMediaTypeError(i18n.t('errors.api.upload.pleaseSelectValidImage')));
+            }
+
+            return options;
         }
 
-        // Check if the file is valid
-        if (!utils.checkFileIsValid(options.uploadimage, config.uploads.contentTypes, config.uploads.extensions)) {
-            return Promise.reject(new errors.UnsupportedMediaTypeError(i18n.t('errors.api.upload.pleaseSelectValidImage')));
+        function storeUpload(options) {
+            var store = storage.getStorage(),
+                filepath = options.data.uploadimage.path;
+
+            return store.save(options.data.uploadimage).then(function (url) {
```

Comment:
`options.data`? Pretty sure this is busted and the integration tests for this api method are junk since they stub out the entire storage module.

tl;dr :sob: